pkg/cmd/extension/push.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extension
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/containerd/reference/docker"
+	"github.com/spf13/cobra"
+
+	wasmimage2 "github.com/tetratelabs/getenvoy/pkg/wasm"
+)
+
+// cmdOpts represents configuration options of the `push` command.
+type cmdOpts struct {
+	wasmFile string
+	imageRef docker.Named
+
+	allowInsecure, plainHTTP bool
+}
+
+func validateWasmFile(wasmFile string, opts *cmdOpts) error {
+	if wasmFile == "" {
+		return fmt.Errorf("WASM file empty")
+	}
+	info, err := os.Stat(wasmFile)
+	if err != nil {
+		return fmt.Errorf("invalid WASM file %q: %w", wasmFile, err)
+	}
+	if info.IsDir() {
+		return fmt.Errorf("WASM file argument was a directory %q", wasmFile)
+	}
+	// Use absolute path because Envoy doesn't run in this directory and may be configured to hot-reload the VM.
+	absoluteWasmFile, err := filepath.Abs(wasmFile)
+	if err != nil {
+		return err
+	}
+	opts.wasmFile = absoluteWasmFile
+	return nil
+}
+
+func validateImageReference(imageRef string, opts *cmdOpts) error {
+	if imageRef == "" {
+		return fmt.Errorf("image reference empty")
+	}
+	ref, err := docker.ParseNormalizedNamed(imageRef)
+	if err != nil {
+		return fmt.Errorf("invalid image reference: %w", err)
+	}
+	opts.imageRef = ref
+	return nil
+}
+
+// NewPushCmd returns a command that pushes the given extension.wasm file
+func NewPushCmd() *cobra.Command {
+	opts := &cmdOpts{}
+
+	cmd := &cobra.Command{
+		Use:   "push /path/to/extension.wasm <image-reference>",
+		Short: "Push the WASM extension to the OCI-compliant registry.",
+		Long: `
+Push the WASM extension to the OCI-compliant registry. This command requires to login the target container registry with docker CLI`,
+		Example: `
+  # Push WASM extension to the local docker registry.
+  getenvoy extension push /path/to/extension.wasm localhost:5000/test/image-name:tag`,
```

Comment:
wonder if the order $path $ref vs $ref $path

![Screen Shot 2021-05-07 at 8 34 20 AM](https://user-images.githubusercontent.com/64215/117381974-f5509000-af0f-11eb-9542-479625672a60.png)

Reply:
@mathetake does embedding oras still make sense? push/pull is basically a wrapper over oras which has more options likely needed for pushing to other registries..

Reply:
maybe in a later PR we can decide to keep or axe this. In an example, we can show oras alternatives to push/pull for wasm https://github.com/deislabs/orasOpensearch